rpc_new/provider/base.ts

```diff
@@ -0,0 +1,71 @@
+import * as msg from "../messages.ts";
+import * as U from "../util.ts";
+import { ProviderCloseError, ProviderHandlerError, ProviderSendError } from "./errors.ts";
+
+/**
+ * @param discoveryValue the value with which to discover the given chain
+ * @param listener the callback to which messages and errors should be applied
+ */
+export type Provider<DiscoveryValue, SendErrorData, HandlerErrorData, CloseErrorData> = (
+  discoveryValue: DiscoveryValue,
+  listener: ProviderListener<SendErrorData, HandlerErrorData>,
+) => ProviderRef<CloseErrorData>;
+
+export type ProviderListener<SendErrorData, HandlerErrorData> = U.Listener<
+  | msg.IngressMessage
+  | ProviderSendError<SendErrorData>
+  | ProviderHandlerError<HandlerErrorData>
+>;
+
+export interface ProviderRef<CloseErrorData> {
+  nextId(): string;
+  send(message: msg.EgressMessage): void;
+  release(): Promise<void | ProviderCloseError<CloseErrorData>>;
+}
+
+declare const providerListener_: unique symbol;
+type providerListener_ = typeof providerListener_;
+
+export abstract class ProviderConnection<Inner, SendErrorData, HandlerErrorData> {
```

Comment:
Why is this abstract?

Reply:
It's the lowest-common denominator between providers: a container for cbs, a connection & some cleanup. Any concerns?

Reply:
Since there are no `abstract` methods or properties, it seems like it could just be a regular class. Then you wouldn't need to have empty subclasses of it.